velox/connectors/hive/HiveDataSink.cpp

```diff
@@ -34,33 +155,100 @@ HiveDataSink::HiveDataSink(
     : inputType_(std::move(inputType)),
       insertTableHandle_(std::move(insertTableHandle)),
       connectorQueryCtx_(connectorQueryCtx),
-      writeProtocol_(std::move(writeProtocol)) {
+      writeProtocol_(std::move(writeProtocol)),
+      partitionChannels_(getPartitionChannels(insertTableHandle_)),
+      partitionIdGenerator_(
+          inputType_,
+          partitionChannels_,
+          connectorQueryCtx_->memoryPool()),
+      numPartitions_(0) {
   VELOX_CHECK_NOT_NULL(
       writeProtocol_, "Write protocol could not be nullptr for HiveDataSink.");
 }
 
 std::shared_ptr<ConnectorCommitInfo> HiveDataSink::getConnectorCommitInfo()
     const {
-  return std::make_shared<HiveConnectorCommitInfo>(writerParameters_);
+  std::vector<std::shared_ptr<const HiveWriterParameters>> toCommit;
+  toCommit.reserve(writerParameters_.size());
+
+  for (auto i = 0; i < writerParameters_.size(); i++) {
+    if (writerParameters_[i] != nullptr) {
+      toCommit.push_back(writerParameters_[i]);
+    }
+  }
+  return std::make_shared<HiveConnectorCommitInfo>(toCommit);
 }
 
 void HiveDataSink::appendData(VectorPtr input) {
-  // For the time being the hive data sink supports one file
-  // To extend it we can create a new writer for every
-  // partition
-  if (writers_.empty()) {
-    writers_.emplace_back(createWriter());
+  input_ = std::dynamic_pointer_cast<RowVector>(input);
+  auto numRows = input->size();
+
+  // Writing to unpartitioned table
+  if (partitionChannels_.size() == 0) {
+    if (writers_.empty()) {
+      createWriterForPartition(0);
+    }
+    writers_[0]->write(input_);
+    return;
+  }
+
+  // Writing to partitioned table
+  std::map<uint64_t, uint64_t> idMap =
+      partitionIdGenerator_.run(input_, partitionIds_);
+
+  // If the current input causes the partitionIdGenerator_ to rehash which
+  // changes the IDs of the previous partition values, move the existing writers
+  // for the previous partition values to new indices.
+  if (!idMap.empty() && !writers_.empty()) {
+    moveWriters(idMap);
+  }
+
+  uint64_t maxPartitionId = partitionIdGenerator_.getMaxPartitionId();
+
+  computePartitionRowCountsAndIndices(
+      partitionIds_,
+      maxPartitionId + 1,
+      connectorQueryCtx_->memoryPool(),
+      partitionSizes_,
+      partitionRows_);
+
+  for (column_index_t i = 0; i < input_->childrenSize(); i++) {
+    input_->childAt(i)->loadedVector();
+  }
+
+  for (vector_size_t partitionId = 0; partitionId <= maxPartitionId;
+       partitionId++) {
+    if (partitionSizes_[partitionId] == 0) {
+      continue;
+    }
+
+    if (writers_.size() < (partitionId + 1) || !writers_[partitionId]) {
+      numPartitions_++;
+      VELOX_CHECK_LE(
+          numPartitions_,
+          HiveConfig::maxPartitionsPerWriters(connectorQueryCtx_->config()),
+          "Exceeded limit of {} open writers for partitions",
```

Comment:
The error message don't need to include the limit. VELOX_CHECK_LE will include numPartition_ and limit automatically.

```
VELOX_CHECK_LE(
          numPartitions_,
          HiveConfig::maxPartitionsPerWriters(connectorQueryCtx_->config()),
          "Exceeded limit of open writers for partitions");
```

Do we have a test that triggers this error?

Reply:
Ah, I see. Kind of saw this before.

Since we're going to move the limit check to the ID generator class, I will add a unit test there for the limit check.

Reply:
Thanks. Let's also make sure to use VELOX_USER_CHECK_LE to generate user exception, not system exception.

Reply:
Makes sense to me